azurelinuxagent/ga/update.py

```diff
@@ -430,13 +445,16 @@ def _process_goal_state(self, exthandlers_handler, remote_access_handler):
             self._report_status(exthandlers_handler, incarnation_changed=False)
             return
 
-        if self._upgrade_available(protocol):
+        if self._check_and_download_agent_if_upgrade_available(protocol):
             available_agent = self.get_latest_agent()
+            # Legacy behavior: The current agent can become unavailable and needs to be reverted.
```

Comment:
not quite "the current agent".  This would exit if none of the versions downloaded are available, right?

Reply:
The function `self.get_latest_agent()` will work as you've described above, i.e. it would return None if all of the downloaded versions are blacklisted. 

But, we would not get into this code path unless the `_check_and_download_agent_if_upgrade_available` returns True, and it will only do so if either a new higher version is actually available (and not blacklisted) or the current version becomes in-eligible (blacklisted) - 

```python
# Remove all blacklisted agents here, so we dont consider blacklisted agents for our computation
self._filter_blacklisted_agents()

# Return True if current agent is no longer available or an
            # agent with a higher version number is available
            return not self._is_version_eligible(base_version) \
                   or (len(self.agents) > 0 and self.agents[0].version > base_version)
```

So this code path will only exit when the current agent indeed does become unavailable. 
Hope that clears it up, the code in this area is a bit convoluted. Hopefully when I get to removing the blacklisting logic, we can make it a bit clearer and easier to understand.

Reply:
Sorry, I meant to say the new comment is a little misleading. If there is no latest agent the code reverts to the daemon installed in the image (I think there is a log statement pointing this out)

Reply:
Ahh I see, that's a good point. I'll update it if I have to push another update, else will handle it in the upcoming PRs